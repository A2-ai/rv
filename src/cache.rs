use std::fmt;
use std::fmt::Formatter;
use std::path::PathBuf;

#[derive(Debug, PartialEq, Clone)]
pub enum CacheEntry {
    Existing(PathBuf),
    NotFound(PathBuf),
    Expired(PathBuf),
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum InstallationStatus {
    Source,
    Binary,
    Both,
    Absent,
}

impl InstallationStatus {
    pub fn available(&self) -> bool {
        *self != InstallationStatus::Absent
    }

    pub fn binary_available(&self) -> bool {
        matches!(self, InstallationStatus::Binary | InstallationStatus::Both)
    }

    pub fn source_available(&self) -> bool {
        matches!(self, InstallationStatus::Source | InstallationStatus::Both)
    }
}

impl fmt::Display for InstallationStatus {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let v = match self {
            InstallationStatus::Source => "source",
            InstallationStatus::Binary => "binary",
            InstallationStatus::Both => "source and binary",
            InstallationStatus::Absent => "absent",
        };
        write!(f, "{v}")
    }
}

pub trait Cache {
    /// This will either load the database for that repository or return None if we couldn't find
    /// it or it was expired.
    fn get_package_db_entry(&self, repo_url: &str) -> CacheEntry;

    /// Gets the status of a package coming from a package repository in the cache
    fn get_package_installation_status(
        &self,
        repo_url: &str,
        name: &str,
        version: &str,
    ) -> InstallationStatus;

    fn get_git_installation_status(&self, repo_url: &str, sha: &str) -> InstallationStatus;

    /// Gets the path to where a git repository should be cloned
    fn get_git_clone_path(&self, repo_url: &str) -> PathBuf;

    /// Gets the path to where a tarball package should be downloaded
    fn get_url_download_path(&self, url: &str) -> PathBuf;
}
