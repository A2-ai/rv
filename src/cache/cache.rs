use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct Cache {
    /// The cache directory.
    /// In practice it will be the OS own cache specific directory + `rv`
    root: PathBuf,
}

impl Cache {
    pub fn new() -> Self {

    }
}