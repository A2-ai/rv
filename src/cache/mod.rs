pub mod disk;
pub mod utils;

pub use disk::{DiskCache, InstallationStatus, PackagePaths};
