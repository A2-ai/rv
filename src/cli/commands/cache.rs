use std::path::Path;

use anyhow::Result;

use crate::{cli::{CliContext, RCommandLookup}, CacheInfo};

fn list_dirs(config_file: impl AsRef<Path>, log_enabled: bool) -> Result<CacheInfo> {
    let mut context = CliContext::new(config_file.as_ref(), RCommandLookup::Skip)?;
    if !log_enabled {
        context.show_progress_bar();
    }
    context.load_databases()?;

    Ok(())
}