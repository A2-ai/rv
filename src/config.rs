use std::collections::HashMap;
use std::path::Path;

use serde::Deserialize;

use crate::r_cmd::RCmd;
use crate::version::Version;

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(deny_unknown_fields)]
struct Author {
    name: String,
    email: String,
    #[serde(default)]
    maintainer: bool,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Repository {
    pub alias: String,
    pub url: String,
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(untagged)]
pub enum DependencyKind {
    Simple(String),
    Detailed {
        name: String,
        repository: Option<String>,
        url: Option<String>,
        #[serde(default)]
        install_suggestions: bool,
        #[serde(default)]
        force_source: bool,
    },
}

impl DependencyKind {
    pub fn name(&self) -> &str {
        match self {
            DependencyKind::Simple(s) => s,
            DependencyKind::Detailed { name, .. } => name,
        }
    }

    pub fn repository(&self) -> Option<&str> {
        match self {
            DependencyKind::Simple(_) => None,
            DependencyKind::Detailed { repository, .. } => repository.as_deref(),
        }
    }

    pub fn force_source(&self) -> bool {
        match self {
            DependencyKind::Simple(_) => false,
            DependencyKind::Detailed { force_source, .. } => *force_source,
        }
    }

    pub fn install_suggestions(&self) -> bool {
        match self {
            DependencyKind::Simple(_) => false,
            DependencyKind::Detailed {
                install_suggestions,
                ..
            } => *install_suggestions,
        }
    }
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub(crate) struct Project {
    name: String,
    #[serde(default)]
    description: String,
    version: Option<Version>,
    license: Option<String>,
    #[serde(default)]
    authors: Vec<Author>,
    #[serde(default)]
    keywords: Vec<String>,
    repositories: Vec<Repository>,
    #[serde(default)]
    suggests: Vec<DependencyKind>,
    r_version: Option<Version>,
    #[serde(default)]
    urls: HashMap<String, String>,
    #[serde(default)]
    dependencies: Vec<DependencyKind>,
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct Config {
    pub(crate) project: Project,
}

impl Config {
    // TODO: handle errors later
    pub fn from_file<P: AsRef<Path>>(path: P) -> Self {
        // TODO: use a custom read file method that reports the filepath if it fails
        let content = std::fs::read_to_string(path).expect("TODO: handle error");
        toml::from_str(content.as_str()).expect("TODO: handle error")
    }

    /// Gets the R version we want to use in this project.
    /// This will default to whatever is set in the config if set, otherwise pick the output
    /// from the trait call
    pub fn get_r_version(&self, r_cli: impl RCmd) -> Version {
        if let Some(v) = &self.project.r_version {
            v.clone()
        } else {
            r_cli.version()
        }
    }

    pub fn repositories(&self) -> &[Repository] {
        &self.project.repositories
    }

    pub fn dependencies(&self) -> &[DependencyKind] {
        &self.project.dependencies
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_parse_valid_config_files() {
        let paths = std::fs::read_dir("src/tests/valid_config/").unwrap();
        for path in paths {
            // TODO: later it will return a res.
            let _ = Config::from_file(path.unwrap().path());
        }
    }
}
