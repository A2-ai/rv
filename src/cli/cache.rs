use std::path::PathBuf;
use std::time::SystemTime;

use anyhow::{anyhow, Context, Result};
use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use etcetera::BaseStrategy;
use fs_err as fs;

use crate::cache::InstallationStatus;
use crate::cli::utils::get_os_path;
use crate::system_info::SystemInfo;
use crate::version::Version;
use crate::{Cache, CacheEntry};

/// How long are the package databases cached for
/// Same default value as PKGCACHE_TIMEOUT:
/// https://github.com/r-lib/pkgcache?tab=readme-ov-file#package-environment-variables
const PACKAGE_TIMEOUT: u64 = 60 * 60;
const PACKAGE_TIMEOUT_ENV_VAR_NAME: &str = "PKGCACHE_TIMEOUT";
const PACKAGE_DB_FILENAME: &str = "packages.bin";

fn get_user_cache_dir() -> Option<PathBuf> {
    etcetera::base_strategy::choose_base_strategy()
        .ok()
        .map(|dirs| dirs.cache_dir().join("rv"))
}

#[inline]
fn get_packages_timeout() -> u64 {
    if let Ok(v) = std::env::var(PACKAGE_TIMEOUT_ENV_VAR_NAME) {
        if let Ok(v2) = v.parse() {
            v2
        } else {
            // If the variable doesn't parse into a valid number, return the default one
            PACKAGE_TIMEOUT
        }
    } else {
        PACKAGE_TIMEOUT
    }
}

/// Just a basic base64 without padding
#[inline]
fn encode_repository_url(url: &str) -> String {
    STANDARD_NO_PAD.encode(url)
}

#[derive(Debug, Clone)]
pub struct PackagePaths {
    pub binary: PathBuf,
    pub source: PathBuf,
}

/// This cache doesn't load anything, it just gets paths to cached objects.
/// Cache freshness is checked when requesting a path and is only a concern for package databases.
#[derive(Debug, Clone)]
pub struct DiskCache {
    /// The cache root directory.
    /// In practice it will be the OS own cache specific directory + `rv`
    root: PathBuf,
    /// R version stored as [major, minor]
    pub r_version: [u32; 2],
    /// The current execution system info: OS, version etc.
    /// Needed for binaries
    pub system_info: SystemInfo,
    /// How long the compiled databases are considered fresh for, in seconds
    /// Defaults to 3600s (1 hour)
    packages_timeout: u64,
    // TODO: check if it's worth keeping a hashmap of repo_url -> encoded
    // TODO: or if the overhead is the same as base64 directly
}

impl DiskCache {
    /// Instantiate our cache abstraction.
    pub fn new(r_version: &Version, system_info: SystemInfo) -> Result<Self> {
        let root =
            get_user_cache_dir().ok_or_else(|| anyhow!("Could not get user cache directory"))?;
        fs::create_dir_all(&root)?;
        cachedir::ensure_tag(&root).context("Failed to create CACHEDIR.TAG")?;

        Ok(Self {
            root,
            system_info,
            r_version: r_version.major_minor(),
            packages_timeout: get_packages_timeout(),
        })
    }

    /// PACKAGES databases as well as binary packages are dependent on the OS and R version
    fn get_repo_root_binary_dir(&self, repo_url: &str) -> PathBuf {
        let encoded = encode_repository_url(repo_url);
        self.root
            .join(&encoded)
            .join(get_os_path(&self.system_info, self.r_version))
    }

    /// A database contains both source and binary PACKAGE data
    /// Therefore the path to the db file is dependent on the system info and R version
    /// In practice it looks like: `CACHE_DIR/rv/{os}/{distrib?}/{arch?}/r_maj.r_min/packages.bin`
    fn get_package_db_path(&self, repo_url: &str) -> PathBuf {
        let base_path = self.get_repo_root_binary_dir(repo_url);
        base_path.join(PACKAGE_DB_FILENAME)
    }

    /// Gets the folder where a binary package would be located.
    /// The folder may or may not exist depending on whether it's in the cache
    pub fn get_binary_package_path(&self, repo_url: &str, name: &str, version: &str) -> PathBuf {
        self.get_repo_root_binary_dir(repo_url)
            .join(name)
            .join(version)
    }

    /// Gets the folder where a source tarball would be located
    /// The folder may or may not exist depending on whether it's in the cache
    pub fn get_source_package_path(&self, repo_url: &str, name: &str, version: &str) -> PathBuf {
        let encoded = encode_repository_url(repo_url);
        self.root.join(encoded).join("src").join(name).join(version)
    }

    pub fn get_package_paths(&self, repo_url: &str, name: &str, version: &str) -> PackagePaths {
        PackagePaths {
            source: self.get_source_package_path(repo_url, name, version),
            binary: self.get_binary_package_path(repo_url, name, version),
        }
    }
}

impl Cache for DiskCache {
    fn get_package_db_entry(&self, repo_url: &str) -> CacheEntry {
        let path = self.get_package_db_path(repo_url);
        if path.exists() {
            let created = path
                .metadata()
                .expect("to work")
                .created()
                .expect("to have a creation time");
            let now = SystemTime::now();

            return if now.duration_since(created).unwrap_or_default().as_secs()
                > self.packages_timeout
            {
                CacheEntry::Expired(path)
            } else {
                CacheEntry::Existing(path)
            };
        }

        CacheEntry::NotFound(path)
    }

    fn get_package_installation_status(
        &self,
        repo_url: &str,
        name: &str,
        version: &str,
    ) -> InstallationStatus {
        let source_present = self
            .get_source_package_path(repo_url, name, version)
            .join(name)
            .is_dir();
        let binary_present = self
            .get_binary_package_path(repo_url, name, version)
            .join(name)
            .is_dir();

        match (source_present, binary_present) {
            (true, true) => InstallationStatus::Both,
            (true, false) => InstallationStatus::Source,
            (false, true) => InstallationStatus::Binary,
            (false, false) => InstallationStatus::Absent,
        }
    }

    fn get_git_clone_path(&self, git_url: &str) -> PathBuf {
        let encoded = encode_repository_url(git_url);
        self.root.join("git").join(encoded)
    }
}
