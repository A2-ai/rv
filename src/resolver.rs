use crate::Cache;
use std::collections::{HashMap, HashSet, VecDeque};
use std::fmt;

use crate::cache::InstallationStatus;
use crate::config::DependencyKind;
use crate::lockfile::Lockfile;
use crate::package::PackageType;
use crate::repository::RepositoryDatabase;
use crate::version::{Version, VersionRequirement};

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct ResolvedDependency<'d> {
    pub(crate) name: &'d str,
    pub(crate) version: &'d str,
    pub(crate) repository_url: &'d str,
    pub(crate) dependencies: Vec<&'d str>,
    pub(crate) force_source: bool,
    pub(crate) kind: PackageType,
    pub(crate) installation_status: InstallationStatus,
    pub(crate) path: Option<&'d str>,
    pub(crate) found_in_lockfile: bool,
}

impl<'d> ResolvedDependency<'d> {
    pub fn is_installed(&self) -> bool {
        match self.kind {
            PackageType::Source => self.installation_status.source_available(),
            PackageType::Binary => self.installation_status.binary_available(),
        }
    }
}

impl<'a> fmt::Display for ResolvedDependency<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}={} (from {}, type={}, from_lockfile={}, path='{}')",
            self.name, self.version, self.repository_url, self.kind, self.found_in_lockfile, self.path.unwrap_or("")
        )
    }
}

#[derive(Debug, PartialEq, Clone)]
enum UnresolvedDependencyKind<'d> {
    /// The user provided a dependency that doesn't exist
    Direct,
    /// A package has a dependency not found. It could be nested several times,
    /// we only show the immediate parent which could be an indirect dep as well.
    Indirect(&'d str),
}

#[derive(Debug, PartialEq, Clone)]
pub struct UnresolvedDependency<'d> {
    name: &'d str,
    version_requirement: Option<&'d VersionRequirement>,
    origins: Vec<UnresolvedDependencyKind<'d>>,
}

impl<'a> fmt::Display for UnresolvedDependency<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut origins = Vec::with_capacity(self.origins.len());
        for origin in &self.origins {
            let v = match origin {
                UnresolvedDependencyKind::Direct => "user provided".to_string(),
                UnresolvedDependencyKind::Indirect(parent) => format!("dependency of `{parent}`"),
            };
            origins.push(v);
        }

        write!(
            f,
            "{}{} {}",
            self.name,
            if let Some(l) = self.version_requirement {
                format!(" {l} ")
            } else {
                String::new()
            },
            format!("from: {}", origins.join(", "))
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ResolutionNeeded {
    /// Nothing to do
    None,
    /// We only need to remove those deps from the lockfile + project library
    RemoveOnly(Vec<String>),
    /// We will need to look up the package databases so do a full lookup, preferring
    /// versions already in the lockfile
    Full,
}

#[derive(Debug, PartialEq)]
pub struct Resolver<'d> {
    /// The repositories are stored in the order defined in the config
    /// The last should get priority over previous repositories
    /// If the bool is `true`, this means this repository should only look at sources
    repositories: &'d [(RepositoryDatabase, bool)],
    r_version: &'d Version,
    /// If we have a lockfile for the resolver, we will skip looking at the database for any package
    /// listed in it
    lockfile: Option<&'d Lockfile>,
}

impl<'d> Resolver<'d> {
    pub fn new(
        repositories: &'d [(RepositoryDatabase, bool)],
        r_version: &'d Version,
        lockfile: Option<&'d Lockfile>,
    ) -> Self {
        Self {
            repositories,
            r_version,
            lockfile,
        }
    }

    pub fn set_repositories(&mut self, repositories: &'d [(RepositoryDatabase, bool)]) {
        self.repositories = repositories;
    }

    /// Tries to find all dependencies from the repos, as well as their install status
    pub fn resolve(
        &self,
        dependencies: &'d [DependencyKind],
        cache: &'d impl Cache,
    ) -> (Vec<ResolvedDependency<'d>>, Vec<UnresolvedDependency<'d>>) {
        let mut resolved = Vec::new();
        // We might have the same unresolved dep multiple times.
        let mut unresolved = HashMap::<&str, UnresolvedDependency>::new();
        let mut found = HashSet::with_capacity(dependencies.len() * 10);

        let mut queue: VecDeque<_> = dependencies
            .iter()
            .map(|d| {
                (
                    d.name(),
                    d.repository(),
                    // required version
                    None,
                    d.install_suggestions(),
                    d.force_source(),
                    None,
                )
            })
            .collect();

        while let Some((
            name,
            repository,
            version_requirement,
            install_suggestions,
            pkg_force_source,
            parent,
        )) = queue.pop_front()
        {
            // If we have already found that dependency, skip it
            // TODO: maybe different version req? we can cross that bridge later
            if found.contains(name) {
                continue;
            }

            // Look at lockfile before looking up any repositories
            if let Some(lockfile) = self.lockfile {
                // If we found the package in the lockfile, consider it found and do not look up
                // the repo at all
                if let Some(package) = lockfile.get_package(name, pkg_force_source) {
                    found.insert(name);
                    resolved.push(ResolvedDependency {
                        name: &package.name,
                        version: &package.version,
                        repository_url: package.repo_url().unwrap(),
                        dependencies: package.dependencies.iter().map(|d| d.as_str()).collect(),
                        kind: if pkg_force_source {
                            PackageType::Source
                        } else {
                            PackageType::Binary
                        },
                        force_source: pkg_force_source,
                        installation_status: cache.get_package_installation_status(
                            package.repo_url().unwrap(),
                            &package.name,
                            &package.version,
                        ),
                        path: package.path.as_ref().map(|x| x.as_str()),
                        found_in_lockfile: true,
                    });

                    for d in &package.dependencies {
                        if !found.contains(d.as_str()) {
                            queue.push_back((d.as_str(), None, None, false, false, Some(name)));
                        }
                    }

                    continue;
                }
            }

            for (repo, repo_source_only) in self.repositories {
                if let Some(r) = repository {
                    if repo.name != r {
                        continue;
                    }
                }

                if let Some((package, package_type)) = repo.find_package(
                    name,
                    version_requirement,
                    self.r_version,
                    pkg_force_source || *repo_source_only,
                ) {
                    found.insert(name);
                    let deps = package.dependencies_to_install(install_suggestions);
                    let all_dependencies = deps.direct;
                  
                    resolved.push(ResolvedDependency {
                        name: &package.name,
                        version: &package.version.original,
                        repository_url: &repo.url,
                        dependencies: all_dependencies.iter().map(|d| d.name()).collect(),
                        kind: package_type,
                        force_source: pkg_force_source || *repo_source_only,
                        installation_status: cache.get_package_installation_status(
                            &repo.url,
                            &package.name,
                            &package.version.original,
                        ),
                        path: package.path.as_ref().map(|x| x.as_str()),
                        found_in_lockfile: false,
                    });

                    for d in all_dependencies {
                        if !found.contains(d.name()) {
                            queue.push_back((
                                d.name(),
                                None,
                                d.version_requirement(),
                                false,
                                false,
                                Some(name),
                            ));
                        }
                    }
                    if install_suggestions {
                        // given all the explicitly called out deps are already in the queue
                        // we don't need to worry about if this is a copy of one of those values
                        // since they'll be in front of the queue and will get found/setup for resolution first
                        // since they're suggested deps we also don't want their suggests for now either,
                        // though one day we might need to add a recursive suggests akin to deps = TRUE
                        // for R. This would cover that situation a bit
                        for d in deps.suggests.unwrap() {
                            if !found.contains(d.name()) {
                                queue.push_back((
                                    d.name(),
                                    None,
                                    d.version_requirement(),
                                    false,
                                    false,
                                    Some(name),
                                ));
                            }
                        }
                    } 
                    break;
                }
            }

            if !found.contains(name) {
                let ud_kind = if let Some(p) = parent {
                    UnresolvedDependencyKind::Indirect(p)
                } else {
                    UnresolvedDependencyKind::Direct
                };
                if let Some(ud) = unresolved.get_mut(name) {
                    ud.origins.push(ud_kind);
                } else {
                    unresolved.insert(
                        name,
                        UnresolvedDependency {
                            name,
                            version_requirement,
                            origins: vec![ud_kind],
                        },
                    );
                }
            }
        }

        (resolved, unresolved.into_values().collect())
    }

    // TODO: add tests
    pub fn resolution_needed(&self, dependencies: &'d [DependencyKind]) -> ResolutionNeeded {
        // If we don't have a lockfile, we'll need to lookup everything
        if self.lockfile.is_none() {
            return ResolutionNeeded::Full;
        }
        let lockfile = self.lockfile.unwrap();
        // At this point we need to figure out 2 things:
        // 1. whether we have all the explicit deps and their deps in the lockfile
        // 2. whether we removed some dep in the config file and need to update the lockfile
        //    and just remove those from the lockfile/rv dir
        let lockfile_deps = lockfile.package_names();
        let mut deps_seen: HashSet<&str> = HashSet::new();

        for d in dependencies {
            // TODO: add source (repository url/git) to the param if set since changing that means a new package
            let all_deps = lockfile.get_package_tree(d.name(), d.force_source());
            // If we don't have an explicit dep in the lockfile, we'll need a full resolve
            if all_deps.is_empty() {
                return ResolutionNeeded::Full;
            }
            deps_seen.extend(all_deps.into_iter());
        }

        // Check whether we have things we need to remove or not
        let unneeded_deps = lockfile_deps
            .difference(&deps_seen)
            .map(|d| d.to_string())
            .collect::<Vec<_>>();

        if unneeded_deps.is_empty() {
            ResolutionNeeded::None
        } else {
            ResolutionNeeded::RemoveOnly(unneeded_deps)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;
    use crate::r_cmd::{InstallError, VersionError};
    use crate::repository::RepositoryDatabase;
    use crate::{CacheEntry, RCmd};
    use serde::Deserialize;
    use std::io::Error;
    use std::path::{Path, PathBuf};
    use std::str::FromStr;

    struct FakeCache;

    impl Cache for FakeCache {
        fn get_package_db_entry(&self, _: &str) -> CacheEntry {
            CacheEntry::NotFound(PathBuf::from_str("").unwrap())
        }

        fn get_package_installation_status(&self, _: &str, _: &str, _: &str) -> InstallationStatus {
            InstallationStatus::Absent
        }
    }

    struct FakeRCmd;
    impl RCmd for FakeRCmd {
        fn install(
            &self,
            _: impl AsRef<Path>,
            _: impl AsRef<Path>,
            _: impl AsRef<Path>,
        ) -> Result<String, InstallError> {
            todo!()
        }

        fn check(
            &self,
            _: &Path,
            _: &Path,
            _: Vec<&str>,
            _: Vec<(&str, &str)>,
        ) -> Result<(), Error> {
            todo!()
        }

        fn build(
            &self,
            _: &Path,
            _: &Path,
            _: &Path,
            _: Vec<&str>,
            _: Vec<(&str, &str)>,
        ) -> Result<(), Error> {
            todo!()
        }

        fn version(&self) -> Result<Version, VersionError> {
            todo!()
        }
    }

    #[derive(Debug, Deserialize)]
    struct TestRepo {
        name: String,
        source: Option<String>,
        binary: Option<String>,
        force_source: bool,
    }

    #[derive(Debug, Deserialize)]
    struct TestRepositories {
        repos: Vec<TestRepo>,
    }

    fn extract_test_elements(
        path: &Path,
    ) -> (Config, Version, Vec<(RepositoryDatabase, bool)>, Lockfile) {
        let content = std::fs::read_to_string(path).unwrap();
        let parts: Vec<_> = content.splitn(3, "---").collect();
        let config = Config::from_str(parts[0]).expect("valid config");
        let r_version = config.get_r_version(FakeRCmd {}).unwrap();
        let repositories = if let Ok(data) = toml::from_str::<TestRepositories>(parts[1]) {
            let mut res = Vec::new();
            for r in data.repos {
                let mut repo = RepositoryDatabase::new(&r.name, &format!("http://{}", r.name));
                if let Some(p) = r.source {
                    let path = format!("src/tests/package_files/{p}.PACKAGE");
                    let text = std::fs::read_to_string(&path).unwrap();
                    repo.parse_source(&text);
                }

                if let Some(p) = r.binary {
                    let path = format!("src/tests/package_files/{p}.PACKAGE");
                    let text = std::fs::read_to_string(&path).unwrap();
                    repo.parse_binary(&text, r_version.major_minor());
                }
                res.push((repo, r.force_source));
            }
            res
        } else {
            let mut repo = RepositoryDatabase::new("inline", "");
            repo.parse_source(parts[1]);
            vec![(repo, false)]
        };
        let lockfile = if parts[2].is_empty() {
            Lockfile::new(&r_version.original)
        } else {
            Lockfile::from_str(parts[2]).expect("valid lockfile")
        };

        (config, r_version, repositories, lockfile)
    }

    #[test]
    fn resolving() {
        let paths = std::fs::read_dir("src/tests/resolution/").unwrap();
        for path in paths {
            let p = path.unwrap().path();
            let (config, r_version, repositories, lockfile) = extract_test_elements(&p);
            let resolver = Resolver::new(&repositories, &r_version, Some(&lockfile));
            let (resolved, unresolved) = resolver.resolve(&config.dependencies(), &FakeCache {});
            // let new_lockfile = Lockfile::from_resolved(&r_version.major_minor(), &resolved);
            // println!("{}", new_lockfile.as_toml_string());
            let mut out = String::new();
            for d in resolved {
                out.push_str(&d.to_string());
                out.push_str("\n");
            }

            if !unresolved.is_empty() {
                out.push_str("--- unresolved --- \n");
                for d in unresolved {
                    out.push_str(&d.to_string());
                    out.push_str("\n");
                }
            }
            // Output has been compared with pkgr for the same PACKAGE file
            insta::assert_snapshot!(p.file_name().unwrap().to_string_lossy().to_string(), out);
        }
    }
}
