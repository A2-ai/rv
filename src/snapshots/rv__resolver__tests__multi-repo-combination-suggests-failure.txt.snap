---
source: src/resolver.rs
expression: out
---
slurmtools=0.1.2 (from http://a2-ai-universe, type=source, from_lockfile=false, path='')
brio=1.1.5 (from http://P3M, type=binary, from_lockfile=false, path='')
dplyr=1.1.4 (from http://P3M, type=binary, from_lockfile=false, path='')
fs=1.6.5 (from http://P3M, type=binary, from_lockfile=false, path='')
glue=1.8.0 (from http://P3M, type=binary, from_lockfile=false, path='')
jsonlite=1.8.9 (from http://P3M, type=binary, from_lockfile=false, path='')
processx=3.8.4 (from http://P3M, type=binary, from_lockfile=false, path='')
purrr=1.0.2 (from http://P3M, type=binary, from_lockfile=false, path='')
rlang=1.1.4 (from http://P3M, type=binary, from_lockfile=false, path='')
tibble=3.2.1 (from http://P3M, type=binary, from_lockfile=false, path='')
whisker=0.4.1 (from http://P3M, type=binary, from_lockfile=false, path='')
withr=3.0.2 (from http://P3M, type=binary, from_lockfile=false, path='')
here=1.0.1 (from http://P3M, type=binary, from_lockfile=false, path='')
cli=3.6.3 (from http://P3M, type=binary, from_lockfile=false, path='')
stringi=1.8.4 (from http://P3M, type=binary, from_lockfile=false, path='')
hms=1.1.3 (from http://P3M, type=binary, from_lockfile=false, path='')
lifecycle=1.0.4 (from http://P3M, type=binary, from_lockfile=false, path='')
knitr=1.49 (from http://P3M, type=binary, from_lockfile=false, path='')
rmarkdown=2.29 (from http://P3M, type=binary, from_lockfile=false, path='')
testthat=3.2.2 (from http://P3M, type=binary, from_lockfile=false, path='')
generics=0.1.3 (from http://P3M, type=binary, from_lockfile=false, path='')
magrittr=2.0.3 (from http://P3M, type=binary, from_lockfile=false, path='')
pillar=1.9.0 (from http://P3M, type=binary, from_lockfile=false, path='')
R6=2.5.1 (from http://P3M, type=binary, from_lockfile=false, path='')
tidyselect=1.2.1 (from http://P3M, type=binary, from_lockfile=false, path='')
vctrs=0.6.5 (from http://P3M, type=binary, from_lockfile=false, path='')
ps=1.8.1 (from http://P3M, type=binary, from_lockfile=false, path='')
fansi=1.0.6 (from http://P3M, type=binary, from_lockfile=false, path='')
pkgconfig=2.0.3 (from http://P3M, type=binary, from_lockfile=false, path='')
rprojroot=2.0.4 (from http://P3M, type=binary, from_lockfile=false, path='')
evaluate=1.0.1 (from http://P3M, type=binary, from_lockfile=false, path='')
highr=0.11 (from http://P3M, type=binary, from_lockfile=false, path='')
xfun=0.49 (from http://P3M, type=binary, from_lockfile=false, path='')
yaml=2.3.10 (from http://P3M, type=binary, from_lockfile=false, path='')
bslib=0.8.0 (from http://P3M, type=binary, from_lockfile=false, path='')
fontawesome=0.5.3 (from http://P3M, type=binary, from_lockfile=false, path='')
htmltools=0.5.8.1 (from http://P3M, type=binary, from_lockfile=false, path='')
jquerylib=0.1.4 (from http://P3M, type=binary, from_lockfile=false, path='')
tinytex=0.54 (from http://P3M, type=binary, from_lockfile=false, path='')
callr=3.7.6 (from http://P3M, type=binary, from_lockfile=false, path='')
desc=1.4.3 (from http://P3M, type=binary, from_lockfile=false, path='')
digest=0.6.37 (from http://P3M, type=binary, from_lockfile=false, path='')
pkgload=1.4.0 (from http://P3M, type=binary, from_lockfile=false, path='')
praise=1.0.0 (from http://P3M, type=binary, from_lockfile=false, path='')
waldo=0.6.1 (from http://P3M, type=binary, from_lockfile=false, path='')
utf8=1.2.4 (from http://P3M, type=binary, from_lockfile=false, path='')
base64enc=0.1-3 (from http://P3M, type=binary, from_lockfile=false, path='')
cachem=1.1.0 (from http://P3M, type=binary, from_lockfile=false, path='')
fastmap=1.2.0 (from http://P3M, type=binary, from_lockfile=false, path='')
memoise=2.0.1 (from http://P3M, type=binary, from_lockfile=false, path='')
mime=0.12 (from http://P3M, type=binary, from_lockfile=false, path='')
sass=0.4.9 (from http://P3M, type=binary, from_lockfile=false, path='')
pkgbuild=1.4.5 (from http://P3M, type=binary, from_lockfile=false, path='')
diffobj=0.3.5 (from http://P3M, type=binary, from_lockfile=false, path='')
rappdirs=0.3.3 (from http://P3M, type=binary, from_lockfile=false, path='')
crayon=1.5.3 (from http://P3M, type=binary, from_lockfile=false, path='')
--- unresolved --- 
bbr from: dependency of `slurmtools`
